google/resource_kms_key_ring.go

```diff
@@ -0,0 +1,142 @@
+	"google.golang.org/api/cloudkms/v1"
+	"log"
+	"regexp"
+	"strings"
+)
+
+func resourceKmsKeyRing() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKmsKeyRingCreate,
+		Read:   resourceKmsKeyRingRead,
+		Delete: resourceKmsKeyRingDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceKmsKeyRingImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"location": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+type kmsKeyRingId struct {
+	Project  string
+	Location string
+	Name     string
+}
+
+func (s *kmsKeyRingId) keyRingId() string {
+	return fmt.Sprintf("projects/%s/locations/%s/keyRings/%s", s.Project, s.Location, s.Name)
+}
+
+func (s *kmsKeyRingId) parentString() string {
+	return fmt.Sprintf("projects/%s/locations/%s", s.Project, s.Location)
+}
+
+func resourceKmsKeyRingCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	keyRingId := &kmsKeyRingId{
+		Project:  project,
+		Location: d.Get("location").(string),
+		Name:     d.Get("name").(string),
+	}
+
+	keyRing, err := config.clientKms.Projects.Locations.KeyRings.Create(keyRingId.parentString(), &cloudkms.KeyRing{}).KeyRingId(keyRingId.Name).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error creating KeyRing: %s", err)
+	}
+
+	log.Printf("[DEBUG] Created KeyRing %s", keyRing.Name)
+
+	d.SetId(keyRing.Name)
+
+	return resourceKmsKeyRingRead(d, meta)
+}
+
+func resourceKmsKeyRingRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	keyRingName := d.Id()
+
+	log.Printf("[DEBUG] Executing read for KMS KeyRing %s", keyRingName)
+
+	_, err := config.clientKms.Projects.Locations.KeyRings.Get(keyRingName).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error reading KeyRing: %s", err)
+	}
+
+	return nil
+}
+
+/*
+	Because KMS KeyRing resources cannot be deleted on GCP, we are only going to remove it from state.
+	Re-creation of this resource through Terraform will produce an error.
+*/
+
+func resourceKmsKeyRingDelete(d *schema.ResourceData, meta interface{}) error {
+	keyRingName := d.Id()
+
+	log.Printf("[WARNING] KMS KeyRing resources cannot be deleted from GCP. This KeyRing %s will be removed from Terraform state, but will still be present on the server.", keyRingName)
+
+	d.SetId("")
+
+	return nil
+}
+
+func parseKmsKeyRingId(id string) (*kmsKeyRingId, error) {
+	keyRingIdRegex := regexp.MustCompile("projects/(.+)/locations/(.+)/keyRings/(.+)$")
+
+	if !keyRingIdRegex.MatchString(id) {
+		return nil, fmt.Errorf("Invalid KeyRing id format, expecting projects/{projectId}/locations/{locationId}/keyRings/{keyRingName}")
```

Comment:
What is the normal pattern for IDs when importing?  I figured I would make it as similar to the REST API as possible, which puts the literal strings `project`, `location`, and `keyRing` in the ID.  Are these normally included in other resources?

Reply:
eh not usually. See https://www.terraform.io/docs/providers/google/r/compute_instance_group.html#import for an example that uses multiple elements in the id.
There's some pretty good discussion on what the best import format is at https://github.com/terraform-providers/terraform-provider-google/pull/378. They ended up with a solution that allows for multiple different ways of importing it, which is pretty neat but not strictly necessary. I'll let you figure out which approach you think works best for KeyRings.

Reply:
Okay, I switched it to `{projectId}/{location}/{keyRing}`